// @sudoku/stores/game.js
import { invalidCells, userGrid } from './grid';
import { SUDOKU_SIZE } from '@sudoku/constants';
import { derived, writable } from 'svelte/store';

export const gamePaused = writable(true);

// Undo/Redo history management
let history = [[]];
let historyIndex = 0;

// Register history callback with userGrid
userGrid.setHistoryCallback((state) => {
	pushHistory(state);
});

export const canUndo = derived(gamePaused, () => historyIndex > 0);
export const canRedo = derived(gamePaused, () => historyIndex < history.length - 1);

export function pushHistory(state) {
	// Remove any redo history when making a new move
	history = history.slice(0, historyIndex + 1);
	history.push(JSON.parse(JSON.stringify(state)));
	historyIndex++;
}

export function undo() {
	if (historyIndex > 0) {
		historyIndex--;
		userGrid.restore(history[historyIndex]);
	}
}

export function redo() {
	if (historyIndex < history.length - 1) {
		historyIndex++;
		userGrid.restore(history[historyIndex]);
	}
}

export const gameWon = derived(
	[userGrid, invalidCells],
	([$userGrid, $invalidCells]) => {
		for (let row = 0; row < SUDOKU_SIZE; row++) {
			for (let col = 0; col < SUDOKU_SIZE; col++) {
				if ($userGrid[row][col] === 0) return false;
			}
		}
		// No empty fields...

		return ($invalidCells.length === 0);
	},
	false,
);

export function pause() {
	gamePaused.set(true);
}

export function resume() {
	gamePaused.set(false);
}
